### ISSUE TITLE
Prepare 17.0.0

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#7396]
This is an umbrella issue to capture some of the decisions we made while working on 16.0.0:

- [ ] Remove CommonJs Node.js API (and update our pure ESM deps)
- [ ] Remove deprecated `output` property in Node.js API result object

I'll create issues for these, but we'll likely want to wait at least a few months before touching them.

In the meantime, we can share any thoughts on what may be, if people have time, good things to look at in `16.x`.

It'd be amazing to see:

- more rules that help people write modern CSS consistently
- a built-in resolver so that people can lint in files in context

I think rules like [`color-format-notation`](https://stylelint.io/user-guide/rules/color-function-notation/) help our users write modern CSS code. I'm seeing code in the wild using modern CSS features like:

- `@container`
- `@layer`
- `clamp()` and other math comparison functions
- `has()`
- logical properties & values
- Nesting
- `oklch()`
- relative colours

For example:

```css
:root {
  --dk-color-black-20: oklch(20% 0 0deg);
  --dk-color-black-20-a-10: oklch(from var(--dk-color-black-20) l c h / 10%);
}

@layer defaults {
  :where(body) {
    color: var(--dk-color-black-20);
  }

  :where(a) {
    &:any-link {
      cursor: pointer;
    }

    &:hover {
      text-decoration: none;
    }
  }
}

parent-component {
  margin-inline: auto;

  &:has(child-component) {
    container: child-component / inline-size;
    display: grid;
  }
}

child-component {
  padding: clamp(1rem, 1rem + 1cqi, 2rem);

  @container parent-component (width > 40rem) {
    display: flex;
  }
}
```

There are already a [couple](https://github.com/stylelint/stylelint/issues/5430) of [open](https://github.com/stylelint/stylelint/issues/6421) issues around these, and I will try to make some time to revisit them. And I'll think about what new rules we can create to help people consistently use these modern features.

The built-in resolver already has an issue (https://github.com/stylelint/stylelint/issues/4088). It's a biggy, though, and I've added a "needs help" label because it's likely non-trivial. It'd be great to have as it'd open up linting CSS in the context of a whole CSS codebase. 

What are the things that you'd like? We can broadly discuss them here, or you can create `needs: discussion` issues for them.

**Comment by jeddy3** [comment]
@ybiquitous Just saw your https://github.com/stylelint/stylelint/issues/7395 issue. What coincidently timing 😆

**Comment by ybiquitous** [comment]
> What coincidental timing 😆

Definitely! 😄 

And I agree with consuming our resources for such modern syntaxes or constructs. 👍🏼

**Comment by Mouvedia** [comment]
It will probably require to compile an ESM migration state list of the top 10 plugins/formatters/syntaxes that we don't maintain.
Not because it won't be done in time, but to give developers the confidence that the ecosystem is ready for this transition.

related: #2498

**Comment by ybiquitous** [comment]
> It will probably require to compile an ESM migration state list of the top 10 plugins/formatters/syntaxes that we don't maintain.

This idea sounds good to me, although our test suite may be more unstable.

**Comment by ybiquitous** [comment]
Added #7556 to the task list.

**Comment by silverwind** [comment]
Are ESM shared configs possible yet? The docs currently say "the module's main file must be a valid JSON configuration", and `main` is the CJS entry point of a module, so I fear the answer is a "no".

**Comment by ybiquitous** [comment]
@silverwind Nice catch. I think the doc you meant is below:

https://github.com/stylelint/stylelint/blob/2d74e26c93b69793d7375d25956a218818ee2601/docs/user-guide/configure.md?plain=1#L286-L290

Indeed, since we have used dynamic [`import()`](https://nodejs.org/api/esm.html#import-expressions) for config modules, Stylelint can now load ESM configs (of course, CommonJS configs are still supported).

<details><summary>Example</summary>
<p>

```jsonc
// package.json
{
  "dependencies": {
    "stylelint": "^16.3.1"
  }
}
```

```js
// stylelint.config.mjs
export default {
  extends: './config.mjs'
};
```

```js
// config.mjs
export default {
  rules: {
    'block-no-empty': true
  }
};
```

Run:

```sh-session
$ npx stylelint -v
16.3.1

$ echo 'a{}' | npx stylelint

<input css c41ZeB>
  1:2  ✖  Unexpected empty block  block-no-empty

✖ 1 problem (1 error, 0 warnings)
```

</p>
</details> 

So, we can add descriptions about ESM to the doc.

**Comment by silverwind** [comment]
Thanks. Yes those are the docs I meant. I will then soon try to migrate my shared config to ESM, thanks!

**Comment by Mouvedia** [comment]
In relation with the [`"node": "^18.17.0 || >=20.5.0"`](https://github.com/stylelint/stylelint/pull/8013#issuecomment-2383728731) change, I think we should add 2 polyfills:

- [`Object.groupBy`](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.object.group-by.js)
- [`Set.prototype.union()`](https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.set.union.v2.js)

The reasons being that:

1. both will require us to drop node 20—which will probably happen in version 19 of Stylelint
2. both have uses in the _current_ codebase

**Comment by ybiquitous** [comment]
Node.js 20.19.0 has supported `require(ESM)` by default! 🎉 
See https://nodejs.org/en/blog/release/v20.19.0

So, I believe we can convert all files to ESM by updating `engines.node` like this:

```diff
   "engines": {
-    "node": ">=18.12.0"
+    "node": ">=20.19.0"
   }
```

**Comment by jeddy3** [comment]
> Node.js 20.19.0 has supported require(ESM) by default! 🎉

Fantastic. It'd be great to remove our build step. It's served us well these past couple of years, though!

We can consider working on `17.0.0` after the end of this month when Node.js 18 reaches EOL.

As an aside, I've been slowly working through and unblocking the issues related to new rules listed in https://github.com/stylelint/stylelint/issues/7396#issue-2035832595. We've added quite of a few of them recently, which is worth celebrating 🎉 .

**Comment by jeddy3** [comment]
Node.js 18 has reached end of life. Shall we start on `17.0.0`? I don't anticipate that we'll need a long-lived branch like we did for `16.0.0`.

It was interesting to read [ESLint's 9.0.0 retrospective](https://eslint.org/blog/2025/05/eslint-v9.0.0-retrospective/):

> Too many breaking changes: Looking back, bundling the configuration system change with the rule API change created too many disruptions. In the future, we will prioritize smaller, more frequent major releases instead of larger, less frequent ones.

With [`require(ESM)` being backported to Node.js `20.19.0`](https://github.com/stylelint/stylelint/issues/7396#issuecomment-2778421740), it doesn't feel like the [task list above](https://github.com/stylelint/stylelint/issues/7396#issue-2035832595) will be too disruptive. I was able to use a local copy of Stylelint that only exported ESM on `20.19.0` with a simple change:

```diff js
- const stylelint  = require("stylelint");
+ const { default: stylelint } = require("stylelint");
```

Am I right in thinking that the VS Code extension [uses `output`](https://github.com/stylelint/vscode-stylelint/blob/d87a95f6dcda3f2fd6c92f6d4126046fcbd11703/src/utils/documents/get-fixes.ts#L25) and we'll need to update it for https://github.com/stylelint/stylelint/issues/7397, but it should be a straightforward change?

I haven't grokked the impact, if any, on users of https://github.com/stylelint/stylelint/issues/8126, though. @romainmenke If you have a moment, can you give a quick assessment?

Apart from those, the changes don't feel disruptive to our users. Dropping support for Node.js 18 will allow us to:

- update our ESM-only dependencies, ensuring we have the latest and most secure versions
- remove the CJS build step, making the codebase easier to maintain

So I think it's well worth doing.

Are there any other changes people would like to see made in `17.0.0`?

**Comment by ybiquitous** [comment]
@firefoxic In theory, blaming may not be a problem even if renaming `.mjs` to `.js` files.
However, in this case, we have to delete `.cjs` at the same time. So, I think it makes sense to avoid risky changes.

Just removing `.cjs` files would never be a problem. Also, we could rename `.mjs` to `.js` in 18.0.0 more safely after we confirm the migration to pure ESM works well.

**Comment by romainmenke** [comment]
> I haven't grokked the impact, if any, on users of https://github.com/stylelint/stylelint/issues/8126, though.

It would be nice if this is green again before we ship 17.0.0: https://github.com/stylelint/stylelint-ecosystem-tester

I am not overly concerned about `8126` as the upgrade path is fairly simple and doesn't require plugin maintainers to do substantial refactors. When they can't or don't want to compute a specific and unambiguous problem range they can provide `endIndex: index + 1`.

It's been a while since we've added new deprecations, so I think that this is a good time to plan 17.0.0 

------

Should we also consider removing `context.fix` in this release?
I don't see it listed here.

**Comment by Mouvedia** [comment]
> context.fix in this release?

That seems like too much for that major. Let's postpone to v18.

**Comment by ybiquitous** [comment]
> It would be nice if this is green again before we ship 17.0.0: [stylelint/stylelint-ecosystem-tester](https://github.com/stylelint/stylelint-ecosystem-tester)

I also hope so, but that may be difficult since the failed packages seem to be unmaintained so much recently: 😓 
- https://github.com/stylus/stylelint-stylus
- https://github.com/stylelint-stylistic/stylelint-stylistic

**Comment by ybiquitous** [comment]
> I am not overly concerned about `8126` as the upgrade path is fairly simple and doesn't require plugin maintainers to do substantial refactors. (...)
>
> It's been a while since we've added new deprecations, so I think that this is a good time to plan 17.0.0

Agree 👍🏼

**Comment by ybiquitous** [comment]
> > context.fix in this release?
> 
> That seems like too much for that major. Let's postpone to v18.

Agree 👍🏼

**Comment by jeddy3** [comment]
A few of our dependencies have dropped support for Node.js 18 or gone ESM-only. Now seems like a good time to create a `v17` branch and start prepping `17.0.0`. I've started the ball rolling with [starting to prepare](https://github.com/stylelint/vscode-stylelint/issues/749) the VS Code extension.

**Comment by romainmenke** [comment]
I'v created a `v17` branch and moved over https://github.com/stylelint/stylelint/pull/8217 to get things started

**Comment by jeddy3** [comment]
With https://github.com/stylelint/stylelint/pull/8859 merged, we can now pick up the other issues and tasks in https://github.com/stylelint/stylelint/issues/7396#issue-2035832595. I've created dedicated issues for some more of the additional task items. Please jump in if you've time. Just change the relevant issue label to `status wip`. 🚀

**Comment by ybiquitous** [comment]
Just for your information, we can see how some packages in the Stylelint ecosystem currently behave with the `v17` branch here:
https://github.com/stylelint/stylelint-ecosystem-tester?tab=readme-ov-file#packages

Ref:
- https://github.com/stylelint/stylelint-ecosystem-tester/pull/98

**Comment by Mouvedia** [comment]
[`stylelintrc.json`](https://github.com/stylelint/stylelint-demo/blob/main/src/schema/stylelintrc.json) will also have to be updated. It should mostly be straightforward.

At least 2 options are being removed in v17.
i.e. `checkContextFunctionalPseudoClasses` and `resolveNestedSelectors`
In that case the `deprecated` annotation keyword won't be appropriate. Since we ought to support v16 alongside v17 for a short while, they won't be removed from the schema for now.

**Comment by jeddy3** [comment]
We've completed the [`17.0.0` task list](https://github.com/stylelint/stylelint/issues/7396#issue-2035832595) changes to this repo. Well done, everyone!

I've opened PRs to prepare our shared configs, demo and website.

We'll want to release the [rearchitected version (`v2`) of the VS Code extension](https://github.com/stylelint/vscode-stylelint/pull/762) first because it'll be compatible with `17.0.0`.  If you have time, testing it would be very helpful!:

1. Clone the [`vscode-stylelint` repo](https://github.com/stylelint/vscode-stylelint).
2. Switch to the `v2` branch.
3. Install deps: `npm ci`.
4. Build extension: `npm run build-bundle`.
5. Use the "Developer: Install Extension from Location..." command: <img width="601" height="62" alt="Image" src="https://github.com/user-attachments/assets/3d1e2844-8aef-415c-9b44-13bd9d1d08ba" />
6. Select the `vscode-stylelint` directory.

You should see a "Local Extension" tag for Stylelint: 
<img width="337" height="64" alt="Image" src="https://github.com/user-attachments/assets/2288a368-3c55-4ca0-9ff5-47e837465caf" />

---

Testing it with `stylelint@16.x` and the `v17` branch (along with the corresponding shared config branch) in your projects would be great:

```shell
npm i -D stylelint/stylelint#v17 stylelint/stylelint-config-standard#prepare-next
```

Along with testing `v17` in your plugins.

Once we've done a round of testing and addressed any bugs that surface, we can release `2.0.0` of the extension and then `17.0.0` of Stylelint.

**Comment by jeddy3** [comment]
I've been [helping out downstream in the SCSS community](https://github.com/stylelint-scss/stylelint-scss/issues/1215) to prepare their packages for Stylelint 17, since our changes to the rules are most disruptive to them. I've suggested we handle the release of the packages this time so we can mention them in our migration guide.

I think we've done our best to strike a good balance between modernising (our support for CSS and our codebase) and minimising disruption to our users, and we're about ready to release. I should be available on the 1st of January to do so.

I've created a release issue: https://github.com/stylelint/stylelint/issues/8943

The last remaining dependency is v2 of the extension: https://github.com/stylelint/vscode-stylelint/issues/751

**Comment by Mouvedia** [comment]
> and minimising disruption to our users

- [x] stylelint-scss
- [x] stylelint-config-recommended-scss
- [x] stylelint-config-standard-scss

Should we at least open issues/PRs on the repertories of these packages?

- @shopify/stylelint-plugin
- @primer/stylelint-config
- stylelint-config-css-modules
- stylelint-config-idiomatic-order
- stylelint-config-property-sort-order-smacss
- stylelint-config-sass-guidelines
- stylelint-config-twbs-bootstrap
- stylelint-config-wikimedia
- stylelint-config-suitcss
- stylelint-config-xo
- stylelint-config-xo-scss
- stylelint-declaration-block-no-ignored-properties
- stylelint-declaration-strict-value
- stylelint-high-performance-animation
- stylelint-less
- stylelint-react-native
- stylelint-checkstyle-reporter

cf https://github.com/stylelint/stylelint-ecosystem-tester

IMHO [`stylelint-less`](https://github.com/stylelint-less/stylelint-less) should at least be fixed before the release.

**Comment by jeddy3** [comment]
I think we've done a good job balancing the time we have as maintainers against shipping something. Typically, after a major release, the broader ecosystem catches up after a couple of weeks.  However, if you have time to open more PRs downstream before we ship, then that'd be fantastic.

**Comment by romainmenke** [comment]
I've also gone over a couple of the downstream packages and most still need to migrate to esm.
This isn't technically difficult, but often opinionated.

For stylelint-less see: https://github.com/stylelint-less/stylelint-less/pull/92

All packages under `@csstools` have already been fixed.

**Comment by Mouvedia** [comment]
> Typically, after a major release, the broader ecosystem catches up after a couple of weeks.

OK.

> This isn't technically difficult, but **often opinionated**.

Right, so we can't assume anything.

> For stylelint-less see: https://github.com/stylelint-less/stylelint-less/pull/92

Thanks.

**Comment by Mouvedia** [comment]
Do need a `next` like we did for v16?
16 had 4 pre-releases
